driver/random.hpp

```diff
@@ -29,4 +38,53 @@ inline T RAN_GEN(T A, T B)
     return r;
 }
 
+template <typename T>
+inline T RAN_SUBNORM(int fraction_msb_bit = 0, int fraction_lsb_bit = 0)
+{
+    (void)fraction_msb_bit;
+    (void)fraction_lsb_bit;
+    return static_cast<T>(0);
+}
+
+#define SUBNORM_INITIALIZE_FRACTION(msb, lsb, msb_default, lsb_default) \
+    do                                                                  \
+    {                                                                   \
+        if((msb) == 0 && (lsb) == 0)                                    \
+        {                                                               \
+            (msb) = (msb_default);                                      \
+            (lsb) = (lsb_default);                                      \
+        }                                                               \
+        assert((msb) <= (msb_default) && (msb) >= (lsb_default));       \
+        assert((lsb) <= (msb_default) && (lsb) >= (lsb_default));       \
+        assert((lsb) <= (msb));                                         \
+    } while(0)
+
+#define SUBNORM_RANDOM_FRACTION(result, msb, lsb)        \
+    do                                                   \
+    {                                                    \
+        int _bits = (msb) - (lsb);                       \
+        if(_bits == 0)                                   \
+            (result) = 0;                                \
+        else                                             \
+            (result) = (rand() % (1 << _bits)) << (lsb); \
```

Comment:
Probably `rand` shouldn't be used here since we've got ridded of it recently. It's slow and not thread safe.
Take a look into this file for the proper random number generator.